Strata/DL/Util/StringGen.lean

```diff
@@ -0,0 +1,120 @@
+  | ⟨d1⟩, ⟨d2⟩ => List.IsSuffix d1 d2
+
+/-- Wrapper around CounterState to allow a prefix -/
+structure StringGenState where
+  cs : Counter.CounterState
+  generated : List (Nat × String)
+
+instance : HasSubset StringGenState where
+  Subset σ₁ σ₂ := σ₁.generated.unzip.2.Subset σ₂.generated.unzip.2
+
+@[simp]
+def StringGenState.emp : StringGenState := { cs := .emp, generated := [] }
+
+def StringGenState.gen (pf : String) (σ : StringGenState) : String × StringGenState :=
+  let (counter, cs) := Counter.genCounter σ.cs
+  let newString : String := (pf ++ "_" ++ toString counter)
+  let newState : StringGenState := { cs, generated := (counter, newString) :: σ.generated }
+  (newString, newState)
+
+def StringGenState.WF (σ : StringGenState)
+  := Counter.WF σ.cs ∧
+    σ.cs.generated = σ.generated.unzip.fst ∧
+    σ.generated.unzip.snd.Nodup ∧
+    -- Q: Is this needed?
+    ∀ c s, (c,s) ∈ σ.generated →
+      String.IsSuffix ("_" ++ toString c) s
+
+theorem String.append_eq_suffix (as bs bs' : String):
+  (as ++ bs = as ++ bs') → bs = bs' := by
+  intros Heq
+  by_cases bs = bs' <;> simp_all
+  next Hne =>
+  have Heq' := String.ext_iff.mp Heq
+  have Hne' : ¬ bs.data = bs'.data := by
+    intros Heq
+    have HH := String.ext_iff.mpr Heq
+    contradiction
+  simp [String.data_append] at *
+  contradiction
+
+theorem String.append_eq_prefix (as as' bs : String):
+  (as ++ bs = as' ++ bs) → as = as' := by
+  intros Heq
+  by_cases as = as' <;> simp_all
+  next Hne =>
+  have Heq' := String.ext_iff.mp Heq
+  have Hne' : ¬ as.data = as'.data := by
+    intros Heq
+    have HH := String.ext_iff.mpr Heq
+    contradiction
+  simp [String.data_append] at *
+  contradiction
+
+theorem List.reverse_injective :
+  List.reverse l₁ = List.reverse l₂ → l₁ = l₂ := List.reverse_inj.mp
+
+theorem String.data_wrap : pf = { data:= pf : String}.data := rfl
+theorem String.data_wrap_eq (a b : String) : a.data = b.data → a = b := String.ext
+
+theorem StringGenState.contains :
+  StringGenState.gen pf σ = (s, σ') →
+  s ∈ σ'.generated.unzip.2 := by
+  intros Hgen
+  simp [gen] at Hgen
+  simp [← Hgen]
+
+theorem StringGenState.subset : gen pf σ = (n, σ') → σ ⊆ σ' := by
+  intros Hgen
+  simp [gen] at Hgen
+  simp [← Hgen, HasSubset.Subset]
+  intros a Hin
+  simp_all
+
+set_option warn.sorry false in
+theorem StringGenState.WFMono :
+  WF σ →
+  gen pf σ = (n, σ') →
+  WF σ' := by
+  intros Hwf Hgen
+  simp [gen] at Hgen
+  cases Hgen with
+  | intro Hl Hr =>
+  simp [← Hr]
+  simp [Hl, WF] at *
+  cases Hwf with
+  | intro Hwfctr Hwf =>
+  cases Hwf with
+  | intro Hwf1 Hwf =>
+  cases Hwf with
+  | intro Hwfnd Hwf12 =>
+  refine ⟨?_, ?_, ?_, ?_⟩
+  . exact Counter.genCounterWFMono Hwfctr rfl
+  . exact
+    Eq.symm
+      (List.reverse_injective
+        (congrArg List.reverse
+          (congrArg (List.cons (Counter.genCounter σ.cs).fst) (id (Eq.symm Hwf1)))))
+  . refine ⟨?_, Hwfnd⟩
+    intros x Hin
+    specialize Hwf12 _ _ Hin
+    sorry
```

Comment:
For each `sorry` in the PR, could you add a comment describing, intuitively, what's being assumed?